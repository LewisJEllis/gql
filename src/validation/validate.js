/* @flow */
/* eslint-disable no-use-before-define */
import { visitInParallel, visit, GraphQLError } from 'graphql';
import { allRules } from './rules';
import type { DocumentNode, GQLError } from '../utils/types';
import GraphQLSchema from '../utils/GraphQLSchema';
import { toGQLError } from '../utils/GQLError';
import { SEVERITY } from '../constants';

export function validate(
  schema: GraphQLSchema,
  ast: DocumentNode,
): Array<GQLError> {
  const context = new ValidationContext(schema, ast);
  const visitors = allRules.map(rule => rule(context));
  visit(ast, visitInParallel(visitors));
  return context.getErrors();
}

// NOTE: keeping the structure of Validation context
// same as graphql-js ValidationContext so we can reuse
// graphql-js validation rule
class ValidationContext {
  _schema: GraphQLSchema;
  _ast: DocumentNode;
  _errors: Array<GQLError> = [];

  constructor(schema: GraphQLSchema, ast: DocumentNode) {
    this._schema = schema;
    this._ast = ast;
  }

  reportError(error: GraphQLError, severity: $Keys<typeof SEVERITY>) {
    this._errors.push(toGQLError(error, severity));
  }

  getErrors(): Array<GQLError> {
    return this._errors;
  }

  getSchema(): GraphQLSchema {
    return this._schema;
  }

  getDocument(): DocumentNode {
    return this._ast;
  }
}
