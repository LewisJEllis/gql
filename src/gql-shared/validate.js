/* @flow */
import { type DocumentNode } from 'graphql/language/ast';
import { type GQLSchema } from './GQLTypes';
import { toGQLError, type GQLError } from './GQLError';
import { type ValidateConfigResolved } from 'gql-config/types';

import { ValidationContext } from 'graphql/validation/validate';
import { TypeInfo } from 'graphql/utilities/TypeInfo';
import {
  visit,
  visitInParallel,
  visitWithTypeInfo,
} from 'graphql/language/visitor';

const makeRuleContext = (context, rule, config) =>
  new Proxy(context, {
    get(target, key) {
      if (key === 'reportError') {
        return error => {
          error.message = `${error.message} (${rule.name})`;
          // $FlowDisableNextLine
          target[key](toGQLError(error, config[rule.name]));
        };
      }
      // $FlowDisableNextLine
      return target[key];
    },
  });

export default function validate(
  schema: GQLSchema,
  ast: DocumentNode,
  validateConfig: ValidateConfigResolved,
): Array<GQLError> {
  const _schema: any = schema;

  if (validateConfig.rules.length === 0) {
    return [];
  }

  const typeInfo = new TypeInfo(_schema);
  const context = new ValidationContext(_schema, ast, typeInfo);
  const visitors = validateConfig.rules.map(rule =>
    rule(makeRuleContext(context, rule, validateConfig.config)),
  );
  visit(ast, visitWithTypeInfo(typeInfo, visitInParallel(visitors)));
  return (context.getErrors(): any);
}
