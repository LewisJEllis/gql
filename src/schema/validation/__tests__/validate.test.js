/* @flow */
import { parse } from 'graphql/language/parser';
import { buildASTSchema } from '../../SchemaBuilder/buildASTSchema';
import { validate } from '../validate';

it('Report unused type definition', () => {
  const ast = parse(`
    type A {
      name: String
    }

    type X {
      value: Int!
    }

    type B {
      name: A
    }
  `);

  const { schema } = buildASTSchema(ast);
  const errors = validate((schema: any), ast);
  expect(errors).toMatchSnapshot();
});

it('throw error if rules packages is unknown', () => {
  const ast = parse(`
    type A {
      name: String
    }

    type X {
      value: Int!
    }

    type B {
      name: A
    }
  `);

  const { schema } = buildASTSchema(ast);
  expect(() => {
    validate((schema: any), ast, { extends: 'some_unknown_package' });
  }).toThrowErrorMatchingSnapshot();
});

