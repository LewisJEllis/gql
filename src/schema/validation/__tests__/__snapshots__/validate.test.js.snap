exports[`test Report unused type definition 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 5,
        "line": 6,
        "path": "GraphQL",
      },
    ],
    "message": "Unused type definition \'X\' (NoUnusedTypeDefinition)",
    "severity": "warn",
  },
  Object {
    "locations": Array [
      Object {
        "column": 5,
        "line": 10,
        "path": "GraphQL",
      },
    ],
    "message": "Unused type definition \'B\' (NoUnusedTypeDefinition)",
    "severity": "warn",
  },
]
`;

exports[`test throw error if rules packages is unknown 1`] = `"unknown validate extends \'some_unknown_package\' available [gql-rules-schema]"`;
